cpp/src/arrow/array/util.cc

```diff
@@ -74,6 +74,186 @@ class ArrayDataWrapper {
   std::shared_ptr<Array>* out_;
 };
 
+class ArrayDataEndianSwapper {
+ public:
+  ArrayDataEndianSwapper(std::shared_ptr<ArrayData>& data, int64_t length)
+      : data_(data), length_(length) {}
+
+  Status SwapType(const DataType& type) { return VisitTypeInline(type, this); }
+
+  Status SwapChildren(std::vector<std::shared_ptr<Field>> child_fields) {
+    int i = 0;
+    for (const auto& child_field : child_fields) {
+      auto orig_data = data_;
+      auto orig_length = length_;
+      data_ = data_->child_data[i++];
+      length_ = data_->length;
+      RETURN_NOT_OK(SwapType(*child_field.get()->type()));
+      length_ = orig_length;
+      data_ = orig_data;
+    }
+    return Status::OK();
+  }
+
+  template <typename T>
+  Status SwapOffset(const T&, int index) {
+    if (data_->buffers[index] == nullptr) {
+      return Status::OK();
+    }
+    using value_type = typename T::c_type;
+    auto buffer = const_cast<value_type*>(
+        reinterpret_cast<const value_type*>(data_->buffers[index]->data()));
+    // offset has one more element rather than data->length
+    int64_t length = length_ + 1;
+    for (int64_t i = 0; i < length; i++) {
+#if ARROW_LITTLE_ENDIAN
+      buffer[i] = BitUtil::FromBigEndian(buffer[i]);
+#else
+      buffer[i] = BitUtil::FromLittleEndian(buffer[i]);
+#endif
+    }
+    return Status::OK();
+  }
+
+  Status SwapSmallOffset(int index = 1) {
+    Int32Type i32;
+    RETURN_NOT_OK(SwapOffset(i32, index));
+    return Status::OK();
+  }
+
+  Status SwapLargeOffset() {
+    Int64Type i64;
+    RETURN_NOT_OK(SwapOffset(i64, 1));
+    return Status::OK();
+  }
```

Comment:
Can we simplify them like the following?

```diff
diff --git a/cpp/src/arrow/array/util.cc b/cpp/src/arrow/array/util.cc
index e2f88016f..6f6d4bc78 100644
--- a/cpp/src/arrow/array/util.cc
+++ b/cpp/src/arrow/array/util.cc
@@ -95,14 +95,13 @@ class ArrayDataEndianSwapper {
     return Status::OK();
   }
 
-  template <typename T>
-  Status SwapOffset(const T&, int index) {
+  template <typename VALUE_TYPE>
+  Status SwapOffset(int index) {
     if (data_->buffers[index] == nullptr) {
       return Status::OK();
     }
-    using value_type = typename T::c_type;
-    auto buffer = const_cast<value_type*>(
-        reinterpret_cast<const value_type*>(data_->buffers[index]->data()));
+    auto buffer = const_cast<VALUE_TYPE*>(
+        reinterpret_cast<const VALUE_TYPE*>(data_->buffers[index]->data()));
     // offset has one more element rather than data->length
     int64_t length = length_ + 1;
     for (int64_t i = 0; i < length; i++) {
@@ -116,15 +115,11 @@ class ArrayDataEndianSwapper {
   }
 
   Status SwapSmallOffset(int index = 1) {
-    Int32Type i32;
-    RETURN_NOT_OK(SwapOffset(i32, index));
-    return Status::OK();
+    return SwapOffset<int32_t>(index);
   }
 
   Status SwapLargeOffset() {
-    Int64Type i64;
-    RETURN_NOT_OK(SwapOffset(i64, 1));
-    return Status::OK();
+    return SwapOffset<int64_t>(1);
   }
 
   template <typename T>
```

Reply:
Thanks for the good suggestion. Done.